File: problem11/julia/matrix_product_cli.jl
include("matrix_product.jl")
using .MatrixProduct

function read_matrix(filename::String)
    open(filename, "r") do io
        return [parse.(Int, split(strip(line))) for line in eachline(io) if !isempty(strip(line))]
    end
end

function print_matrix_with_highlight(matrix::Array{Int,2}, coords::Vector{Tuple{Int,Int}})
    coordset = Set(coords)
    rows, cols = size(matrix)
    for i in 1:rows
        for j in 1:cols
            val = matrix[i, j]
            if (i, j) in coordset
                print("\e[31m", lpad(val, 2, '0'), "\e[0m ")
            else
                print(lpad(val, 2, '0'), " ")
            end
        end
        println()
    end
end

function main()
    filename = length(ARGS) > 0 ? ARGS[1] : "matrix.txt"
    mat = read_matrix(filename)
    matrix = reduce(vcat, [row' for row in mat])
    result, coords = greatest_product(matrix, 4)
    print_matrix_with_highlight(matrix, coords)
    println("Greatest product of four adjacent numbers: $result")
end

main()

File: problem11/perl/matrix_product_cli.pl
#!/usr/bin/env perl
use strict;
use warnings;
use FindBin;
use lib $FindBin::Bin;
require 'matrix_product.pl';
use Term::ANSIColor;
use Scalar::Util qw(looks_like_number);

sub read_matrix {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Failed to read matrix from $filename: $!\n";
    my @matrix;
    while (my $line = <$fh>) {
        chomp $line;
        my @row = grep { looks_like_number($_) } split /\s+/, $line;
        push @matrix, \@row if @row;
    }
    close $fh;
    return \@matrix;
}

sub print_matrix_with_highlight {
    my ($matrix, $coords) = @_;
    my %coord_set = map { join(',', @$_) => 1 } @$coords;
    for my $i (0..$#$matrix) {
        for my $j (0..$#{$matrix->[$i]}) {
            my $val = $matrix->[$i][$j];
            if ($coord_set{"$i,$j"}) {
                print color('red'), sprintf("%02d ", $val), color('reset');
            } else {
                print sprintf("%02d ", $val);
            }
        }
        print "\n";
    }
}

# Main
my $filename = $ARGV[0] // 'matrix.txt';
my $matrix = read_matrix($filename);
my ($result, $coords) = greatest_product($matrix, 4);
print_matrix_with_highlight($matrix, $coords);
print "Greatest product of four adjacent numbers: $result\n";
